torch/fx/experimental/symbolic_shapes.py

```diff
@@ -683,6 +710,39 @@ def codegen_guards(self, placeholders, sources,
 
         from torch._dynamo.source import NegateSource, TensorPropertySource, TensorProperty
 
+        SourceInfo = collections.namedtuple("SourceInfo", ["base", "sources"])
+        SymbolicShapesExpr = collections.namedtuple("SymbolicShapeExpr", ["guard_expr", "call_expr", "fn"])
+        source_name_to_types = {}
+        source_name_to_infos = collections.defaultdict(lambda: SourceInfo(base=None, sources=[]))
+
+        def split_source_ref(source: Source) -> Tuple[str, str]:
+            ref = source_ref(source)
+            if isinstance(source, TensorPropertySource) and source.idx is not None:
+                end_position = -(1 + len(str(source.idx)) + 1)
+                return ref[0:end_position], ref[end_position:]
+            else:
+                return ref, ""
+
+        def flat_var_name(var_name: str) -> str:
+            var_name = var_name.replace(" ", "_")
+            var_name = var_name.replace("-", "_Neg_")
+            var_name = var_name.replace("[", "_LBr_").replace("]", "_RBr_")
+            var_name = var_name.replace("(", "_LPar_").replace(")", "_RPar_")
+            var_name = var_name.replace("'", "_Q_")
+            var_name = var_name.replace(".", "_D_").replace(",", "_C_")
+            return var_name
+
+        def flat_source_ref(source: Source) -> str:
+            name, position = split_source_ref(source)
+            return f"{flat_var_name(name)}{position}"
```

Comment:
I needed these to turn Python expressions into unique variables. The good thing about it is that the variable name is informative (easier to reason about it when debugging). 

A, perhaps, cleaner approach would be to create new variables `varN` for each of them. The bad thing about it is that we lose the original name entirely.

Any thoughts on this?

Reply:
you can use the old name but with a counter that increments on conflict

Reply:
How would that work, for example, in expressions like:

- `getattr(getattr(self.model.decoder.block, '2').layer, '1').EncDecAttention.k.weight.stride()`
    - Currently turned into: `getattr_LPar_getattr_LPar_self_D_model_D_decoder_D_block_C___Q_2_Q__RPar__D_layer_C___Q_1_Q__RPar__D_EncDecAttention_D_k_D_weight_D_stride_LPar__RPar_` 
- `self.model.decoder.final_layer_norm.weight.stride()`
    - Currently turned into: `self_D_model_D_decoder_D_final_layer_norm_D_weight_D_stride_LPar__RPar_`

Reply:
Basically, you can keep using the mangling scheme you do here, but each mangled name needs to remember what expression it corresponds to, and if you have two expressions that happen to mangle to the same string, you need to add an identifier to distinguish them.

Reply:
Hmm... I am a little bit confused.
If it's better to keep using this mangling, then there should be no other problems (i.e. no need for a counter). The only problem was that using `replace` for transforming expressions into identifiers might not be a great idea (depends on what characters the expression contains).

Reply:
OK, sorry, this a little bit of a nuanced discussion.

The problem is your mangler does not guarantee that if expr1 != expr2, then mangle(expr1) != mangle(expr2). The problem is you're not mangling underscores in the source; so I can write `a_b` and this will mangle to the same thing as `a b`. If you ensure mangling is reversible, then there is no problem and we don't need counters.

Reply:
Ah, now I see what you mean. Thanks for the explanation.